compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -454,29 +461,145 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(implicit ctx: Context) {
       }
 
       // Drop unused bindings
-      val (finalBindings, finalExpansion) = dropUnusedDefs(bindingsBuf.toList, expansion1)
+      val matchBindings = reducer.matchBindingsBuf.toList
+      val (finalBindings, finalExpansion) = dropUnusedDefs(bindingsBuf.toList ++ matchBindings, expansion1)
+      val (finalMatchBindings, finalArgBindings) = finalBindings.partition(matchBindings.contains(_))
 
       if (inlinedMethod == defn.Typelevel_error) issueError()
 
       // Take care that only argument bindings go into `bindings`, since positions are
       // different for bindings from arguments and bindings from body.
-      tpd.Inlined(call, finalBindings, finalExpansion)
+      tpd.Inlined(call, finalArgBindings, seq(finalMatchBindings, finalExpansion))
     }
   }
 
   /** A utility object offering methods for rewriting inlined code */
   object reducer {
 
+    /** Additional bindings established by reducing match expressions */
+    val matchBindingsBuf = new mutable.ListBuffer[MemberDef]
+
+    /** An extractor for terms equivalent to `new C(args)`, returning the class `C`,
+     *  a list of bindings, and the arguments `args`. Can see inside blocks and Inlined nodes and can
+     *  follow a reference to an inline value binding to its right hand side.
+     *  @return    optionally, a triple consisting of
+     *             - the class `C`
+     *             - the arguments `args`
+     *             - any bindings that wrap the instance creation
+     *             - whether the instance creation is precomputed or by-name
+     */
+    private object NewInstance {
+      def unapply(tree: Tree)(implicit ctx: Context): Option[(Symbol, List[Tree], List[Tree], Boolean)] = {
+        def unapplyLet(bindings: List[Tree], expr: Tree) =
+          unapply(expr) map {
+            case (cls, reduced, prefix, precomputed) => (cls, reduced, bindings ::: prefix, precomputed)
+          }
+        tree match {
+          case Apply(fn, args) =>
```

Comment:
We don't handle that right now (i.e. we cannot reduce a projection involving  constructor with multiple argument lists). I'd leave that as a TODO if we want to pursue the partial evaluation route. I think this one would not be among the most pressing extensions, and it would be quite tricky to implement.